gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/oauth2/ExternalAuthTokenFilter.groovy

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2016 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.gate.security.oauth2
+
+import groovy.util.logging.Slf4j
+import org.apache.commons.lang3.StringUtils
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Qualifier
+import org.springframework.security.oauth2.client.OAuth2RestOperations
+import org.springframework.security.oauth2.common.DefaultOAuth2AccessToken
+import org.springframework.security.oauth2.common.OAuth2AccessToken
+
+import javax.servlet.Filter
+import javax.servlet.FilterChain
+import javax.servlet.FilterConfig
+import javax.servlet.ServletException
+import javax.servlet.ServletRequest
+import javax.servlet.ServletResponse
+import javax.servlet.http.HttpServletRequest
+
+@Slf4j
+class ExternalAuthTokenFilter implements Filter {
+
+  @Autowired
+  @Qualifier("userInfoRestTemplate")
+  OAuth2RestOperations restTemplate
+
+  @Override
+  void init(FilterConfig filterConfig) throws ServletException {
+
+  }
+
+  @Override
+  void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+    def httpServletRequest = (HttpServletRequest) request
+    String authorizationHeader = httpServletRequest.getHeader("Authorization")
+    if (StringUtils.isNotEmpty(authorizationHeader) && authorizationHeader.startsWith(OAuth2AccessToken.BEARER_TYPE)) {
+      log.debug("Authorization header detected, attempting to extract bearer token.")
+      String token = StringUtils.substringAfterLast(authorizationHeader, " ")
+      if (StringUtils.isNotEmpty(token)) {
+        if (log.traceEnabled) {
+          log.trace("Bearer token extracted: " + token) // Don't turn this on in production!
+        } else if (log.debugEnabled) {
+          log.debug("Bearer token extracted.")
+        }
+        restTemplate.OAuth2ClientContext.accessToken = new DefaultOAuth2AccessToken(token)
```

Comment:
Jake and I spent some time reverse engineering how this statement works with multiple users. Turns out, the OAuth2ClientContext is session scoped, so each user will have their own, and this statement is safe. 

See https://github.com/spring-cloud/spring-cloud-security/blob/66cc52bd46e9f2585733bdedec9d56e49d8516f5/spring-cloud-security/src/main/java/org/springframework/cloud/security/oauth2/client/OAuth2ClientAutoConfiguration.java#L125